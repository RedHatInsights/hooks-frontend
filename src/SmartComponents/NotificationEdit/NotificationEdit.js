import React, { Component } from 'react';
import { Button } from '@patternfly/react-core';

import { bindActionCreators } from 'redux';
import { withRouter, Redirect } from 'react-router-dom';
import Form from 'react-jsonschema-form';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
    fetchEndpoint,
    createEndpoint,
    updateEndpoint,
    newEndpoint,
    fetchFilters,
    fetchApps
} from '../../store/actions';
import {
    Skeleton,
    SkeletonSize
} from '@red-hat-insights/insights-frontend-components';
import registryDecorator from '@red-hat-insights/insights-frontend-components/Utilities/Registry';
import {
    LoadingState,
    NotificationsPage,
    FilterList
} from '../../';

const schema = {
    title: 'Edit Notifications',
    type: 'object',
    required: [ 'name', 'url' ],
    properties: {
        name: { type: 'string', title: 'Name' },
        active: { type: 'boolean', title: 'Active', default: true },
        url: { type: 'string', title: 'URL' }
    }
};

const uiSchema = {
    name: {
        'ui:placeholder': 'New notification endpoint name'
    },
    url: {
        'ui:placeholder': 'https://...'
    }
};

function CustomFieldTemplate(props) {
    const { id, classNames, label, help, required, description, errors, children } = props;
    let allClassNames = classNames.concat([ ' pf-c-form__group' ]);

    return (
        <div className={ allClassNames }>
            <label htmlFor={ id }>{ label } { required ? '*' : null }</label>
            { description }
            { children }
            { errors }
            { help }
        </div>
    );
}

CustomFieldTemplate.propTypes = {
    id: PropTypes.string,
    classNames: PropTypes.string,
    label: PropTypes.string,
    help: PropTypes.object,
    required: PropTypes.bool,
    description: PropTypes.object,
    errors: PropTypes.object,
    children: PropTypes.array
};

@registryDecorator()
export class NotificationEdit extends Component {
    constructor(props) {
        super(props);
        this.filterList = React.createRef();
        this.fetchData();
    }

    formSubmit = (data) => {
        let { active, name, url } = data.formData;
        let filters = [{
            app_ids: this.filterList.current.props.selectedAppEventTypes.appIds,
            event_type_ids: this.filterList.current.props.selectedAppEventTypes.eventTypeIds,
            severity_filters: []
        }];
        let payload = {
            active,
            name,
            url,
            filters
        };

        if (this.props.endpoint) {
            this.props.updateEndpoint(this.props.endpoint.id, payload);
        } else {
            this.props.createEndpoint(payload);
        }

        this.fetchData();
    };

    fetchData = () => {
        let id = this.props.match.params.endpointId;

        if (id) {
            this.props.fetchEndpoint(id);
            this.props.fetchFilters(id);
        }

        this.props.fetchApps();
    }

    selectedAppEventTypes = () => {
        if (this.props.filters && this.props.filters.length > 0) {
            return {
                appIds: this.props.filters.map((filter) =>
                    filter.relationships.apps.data.map((app) => parseInt(app.id))).flat(),
                eventTypeIds: this.props.filters.map((filter) =>
                    filter.relationships.event_types.data.map((eventType) => parseInt(eventType.id))).flat()
            };
        } else {
            return {
                appIds: [],
                eventTypeIds: []
            };
        }
    }

    initialFormData = () => {
        return this.props.endpoint ? {
            name: this.props.endpoint.name,
            url: this.props.endpoint.url,
            active: this.props.endpoint.active
        } : {};
    }

    onCancel = () => {
        this.props.history.push('/list');
    }

    render() {
        const action = this.props.match.params.endpointId ? 'Edit' : 'New';

        if (this.props.endpoint && !this.props.match.params.endpointId) {
            return <Redirect to={ `/edit/${ this.props.endpoint.id }` } />;
        }

        return (
            <NotificationsPage title={ `${ action } Notification` }>
                <LoadingState
                    loading={ this.props.loading }
                    placeholder={ <Skeleton size={ SkeletonSize.sm } /> }>
                    <Form schema={ schema } className="pf-c-form"
                        uiSchema={ uiSchema }
                        formData={ this.initialFormData() }
                        onSubmit={ this.formSubmit }
                        FieldTemplate={ CustomFieldTemplate }>

                        <FilterList ref={ this.filterList }
                            apps={ this.props.apps }
                            selectedAppEventTypes={ this.selectedAppEventTypes() } />

                        <div>
                            <Button type='submit' variant="primary">Submit</Button>
                            <Button onClick={ this.onCancel } variant="secondary">Cancel</Button>
                        </div>
                    </Form>
                </LoadingState>
            </NotificationsPage>
        );
    }
}

NotificationEdit.propTypes = {
    endpointId: PropTypes.number,
    endpoint: PropTypes.object,
    filters: PropTypes.array.isRequired,
    apps: PropTypes.array.isRequired,
    fetchEndpoint: PropTypes.func.isRequired,
    createEndpoint: PropTypes.func.isRequired,
    updateEndpoint: PropTypes.func.isRequired,
    fetchFilters: PropTypes.func.isRequired,
    fetchApps: PropTypes.func.isRequired,
    match: PropTypes.object,
    history: PropTypes.object,
    loading: PropTypes.bool,
    submitting: PropTypes.bool
};

const mapStateToProps = function(state) {
    let { endpoint, loading, submitting } = state.endpoints;
    let { apps, loading: appsLoading } = state.apps;
    let { filters, loading: filtersLoading } = state.filters;

    return {
        endpoint,
        apps,
        filters,
        loading,
        submitting,
        appsLoading,
        filtersLoading
    };
};

const mapDispatchToProps = function (dispatch) {
    return bindActionCreators({
        fetchEndpoint,
        createEndpoint,
        updateEndpoint,
        newEndpoint,
        fetchFilters,
        fetchApps
    }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(NotificationEdit));
