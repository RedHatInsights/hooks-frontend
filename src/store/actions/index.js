import _ from 'lodash';
import ApiClient from 'Utilities/notificationsBackendAPI';

export const FETCH_FILTER    = 'FETCH_FILTER';
export const FETCH_ENDPOINTS = 'FETCH_ENDPOINTS';
export const FETCH_ENDPOINT  = 'FETCH_ENDPOINT';
export const SUBMIT_ENDPOINT = 'SUBMIT_ENDPOINT';
export const DELETE_ENDPOINT = 'DELETE_ENDPOINT';
export const NEW_ENDPOINT    = 'NEW_ENDPOINT';
export const TEST_ENDPOINT   = 'TEST_ENDPOINT';
export const FETCH_APPS      = 'FETCH_APPS';

export const fetchEndpoints = (limit, offset, sortBy = 'name asc') => {
    const params = { limit, offset, sort_by: sortBy };
    const query = _.map(params, (value, key) => value ? `${ key }=${ value }` : '');
    const url = `/endpoints?${ query.join('&') }`;

    return {
        type: FETCH_ENDPOINTS,
        payload: ApiClient.get(url),
        meta: {
            endpoint: url,
            sortBy,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: 'Failed to load endpoints'
                }
            }
        }
    };
};

export const fetchEndpoint = (id) => ({
    type: FETCH_ENDPOINT,
    payload: ApiClient.get(`/endpoints/${ id }`),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: `Failed to load endpoint ${ id }`
            }
        }
    }
});

export const createEndpoint = (data) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.create('/endpoints', { endpoint: data }),
        meta: {
            data,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to create endpoint ${ data.name }`
                },
                fulfilled: {
                    variant: 'success',
                    title: `Endpoint ${ data.name } created`
                }
            }
        }
    };
};

export const updateEndpoint = (id, data) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.update(`/endpoints/${ id }`, { endpoint: data }),
        meta: {
            data,
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to update endpoint ${ data.name }`
                },
                fulfilled: {
                    variant: 'success',
                    title: `Endpoint ${ data.name } updated`
                }
            }
        }
    };
};

export const toggleEndpoint = (id, on) => {
    return {
        type: SUBMIT_ENDPOINT,
        payload: ApiClient.update(`/endpoints/${ id }`, { endpoint: { active: on }}),
        meta: {
            notifications: {
                rejected: {
                    variant: 'danger',
                    title: `Failed to toggle endpoint ${ id }`
                }
            }
        }
    };
};

export const deleteEndpoint = (id, name) => ({
    type: DELETE_ENDPOINT,
    payload: ApiClient.destroy(`/endpoints/${ id }`).then(() => ({ id })),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: `Failed to delete endpoint ${ name }`
            },
            fulfilled: {
                variant: 'success',
                title: `Endpoint ${ name } deleted`
            }
        }
    }
});

export const newEndpoint = () => ({
    type: NEW_ENDPOINT
});

export const testEndpoint = (endpointId) => ({
    type: TEST_ENDPOINT,
    payload: ApiClient.create(`/endpoints/${ endpointId }/test`, {}),
    meta: {
        notifications: {
            rejected: {
                variant: 'warning',
                title: 'Test event delivery failed'
            },
            fulfilled: {
                variant: 'success',
                title: 'Test event delivery successful'
            }
        }
    }
});

export const fetchFilter = (endpointId) => ({
    type: FETCH_FILTER,
    payload: ApiClient.get(`/endpoints/${ endpointId }/filter`),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: 'Failed to load filter'
            }
        }
    }
});

export const fetchApps = () => ({
    type: FETCH_APPS,
    payload: ApiClient.get('/apps'),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: 'Failed to load apps'
            }
        }
    }
});
