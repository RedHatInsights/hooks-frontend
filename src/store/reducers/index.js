import {
    FETCH_FILTERS,
    FETCH_FILTERS_SUCCESS,
    FETCH_FILTERS_FAILURE,
    FETCH_ENDPOINTS,
    FETCH_ENDPOINTS_SUCCESS,
    FETCH_ENDPOINTS_FAILURE,
    FETCH_ENDPOINT_SUCCESS
} from '../actions/index';

const defaultIntialState = {
    loading: false,
    error: null
};

const initialStateFor = function (reducerName) {
    let initState = Object.assign({}, defaultIntialState);
    initState[reducerName] = [];
    return initState;
};

export const endpointReducer = function(state = initialStateFor('endpoints'), action) {
    switch (action.type) {
        case FETCH_ENDPOINTS:
            return {
                ...state,
                loading: true,
                error: null
            };

        case FETCH_ENDPOINTS_SUCCESS:
            return {
                ...state,
                loading: false,
                endpoints: action.payload.endpoints
            };

        case FETCH_ENDPOINT_SUCCESS:
            return {
                ...state,
                loading: false,
                endpoint: action.payload.endpoint
            };

        case FETCH_ENDPOINTS_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error,
                endpoints: []
            };

        default:
            return state;
    }
};

export const filterReducer = function(state = initialStateFor('filters'), action) {
    switch (action.type) {
        case FETCH_FILTERS:
            return {
                ...state,
                loading: true,
                error: null
            };

        case FETCH_FILTERS_SUCCESS:
            return {
                ...state,
                loading: false,
                filters: action.payload.filters
            };

        case FETCH_FILTERS_FAILURE:
            return {
                ...state,
                loading: false,
                error: action.payload.error,
                filters: []
            };

        default:
            return state;
    }
};
