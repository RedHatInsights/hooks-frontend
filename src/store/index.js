import ReducerRegistry from '@red-hat-insights/insights-frontend-components/Utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import { filterReducer, endpointsReducer } from './reducers';
import { appsReducer } from './reducers/AppsReducer';
import { notifications, notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';

let registry;

export function init (...middleware) {
    if (registry) {
        throw new Error('store already initialized');
    }

    registry = new ReducerRegistry({}, [
        promiseMiddleware(),
        notificationsMiddleware(),
        ...middleware
    ]);

    registry.register({
        filter: filterReducer,
        endpoints: endpointsReducer,
        apps: appsReducer,
        notifications
    });

    return registry;
}

export function getStore () {
    return registry.getStore();
}

export function register (...args) {
    return registry.register(...args);
}
