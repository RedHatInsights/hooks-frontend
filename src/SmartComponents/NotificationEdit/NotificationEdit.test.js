import { init } from '../../store';
import logger from 'redux-logger';
import { NotificationEdit } from './NotificationEdit';

const testEndpoint = {
    id: 3,
    name: 'TEST Endpoint #3',
    url: 'http://endpoint3.com',
    active: false,
    filtersCount: 4
};

describe('NotificationEdit', () => {
    const store = init(logger).getStore();
    const fetchEndpoint = jest.fn();
    const createEndpoint = jest.fn();
    const updateEndpoint = jest.fn();
    const fetchFilters = jest.fn();
    const fetchApps = jest.fn();

    const defaultProps = {
        match: {
            params: {
                endpointId: '1'
            }
        },
        fetchEndpoint,
        createEndpoint,
        updateEndpoint,
        fetchFilters,
        fetchApps,
        store
    };

    it('expect to render a Form', () => {
        const wrapper = shallow(
            <NotificationEdit { ...defaultProps } filters={ [] } apps={ [] } />
        );
        expect(wrapper.find('Form').length).toBe(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('takes an endpoint', () => {
        const wrapper = shallow(
            <NotificationEdit { ...defaultProps } endpoint={ testEndpoint } filters={ [] } apps={ [] } />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
