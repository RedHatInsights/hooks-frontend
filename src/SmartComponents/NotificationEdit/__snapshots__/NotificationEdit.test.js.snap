// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationEdit expect to render a Form 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationEdit
    fetchEndpoint={
      [MockFunction] {
        "calls": Array [
          Array [
            "1",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "endpointId": "1",
        },
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <PageHeader>
          <PageHeaderTitle
            title="Edit Notification"
          />
        </PageHeader>,
        <Connect(Main)>
          <Form
            ErrorList={[Function]}
            FieldTemplate={[Function]}
            className="pf-c-form"
            disabled={false}
            formData={Object {}}
            liveValidate={false}
            noHtml5Validate={false}
            noValidate={false}
            onChange={[Function]}
            safeRenderCompletion={false}
            schema={
              Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              }
            }
            uiSchema={Object {}}
          />
        </Connect(Main)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <PageHeaderTitle
            title="Edit Notification"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "title": "Edit Notification",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Form
            ErrorList={[Function]}
            FieldTemplate={[Function]}
            className="pf-c-form"
            disabled={false}
            formData={Object {}}
            liveValidate={false}
            noHtml5Validate={false}
            noValidate={false}
            onChange={[Function]}
            safeRenderCompletion={false}
            schema={
              Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              }
            }
            uiSchema={Object {}}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "ErrorList": [Function],
            "FieldTemplate": [Function],
            "className": "pf-c-form",
            "disabled": false,
            "formData": Object {},
            "liveValidate": false,
            "noHtml5Validate": false,
            "noValidate": false,
            "onChange": [Function],
            "safeRenderCompletion": false,
            "schema": Object {
              "properties": Object {
                "active": Object {
                  "default": true,
                  "title": "Active",
                  "type": "boolean",
                },
                "name": Object {
                  "default": "New notification endpoint name",
                  "title": "Name",
                  "type": "string",
                },
                "url": Object {
                  "default": "https://...",
                  "title": "URL",
                  "type": "string",
                },
              },
              "required": Array [
                "title",
              ],
              "title": "Edit Notifications",
              "type": "object",
            },
            "uiSchema": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <PageHeader>
            <PageHeaderTitle
              title="Edit Notification"
            />
          </PageHeader>,
          <Connect(Main)>
            <Form
              ErrorList={[Function]}
              FieldTemplate={[Function]}
              className="pf-c-form"
              disabled={false}
              formData={Object {}}
              liveValidate={false}
              noHtml5Validate={false}
              noValidate={false}
              onChange={[Function]}
              safeRenderCompletion={false}
              schema={
                Object {
                  "properties": Object {
                    "active": Object {
                      "default": true,
                      "title": "Active",
                      "type": "boolean",
                    },
                    "name": Object {
                      "default": "New notification endpoint name",
                      "title": "Name",
                      "type": "string",
                    },
                    "url": Object {
                      "default": "https://...",
                      "title": "URL",
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "title",
                  ],
                  "title": "Edit Notifications",
                  "type": "object",
                }
              }
              uiSchema={Object {}}
            />
          </Connect(Main)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <PageHeaderTitle
              title="Edit Notification"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "title": "Edit Notification",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Form
              ErrorList={[Function]}
              FieldTemplate={[Function]}
              className="pf-c-form"
              disabled={false}
              formData={Object {}}
              liveValidate={false}
              noHtml5Validate={false}
              noValidate={false}
              onChange={[Function]}
              safeRenderCompletion={false}
              schema={
                Object {
                  "properties": Object {
                    "active": Object {
                      "default": true,
                      "title": "Active",
                      "type": "boolean",
                    },
                    "name": Object {
                      "default": "New notification endpoint name",
                      "title": "Name",
                      "type": "string",
                    },
                    "url": Object {
                      "default": "https://...",
                      "title": "URL",
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "title",
                  ],
                  "title": "Edit Notifications",
                  "type": "object",
                }
              }
              uiSchema={Object {}}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "ErrorList": [Function],
              "FieldTemplate": [Function],
              "className": "pf-c-form",
              "disabled": false,
              "formData": Object {},
              "liveValidate": false,
              "noHtml5Validate": false,
              "noValidate": false,
              "onChange": [Function],
              "safeRenderCompletion": false,
              "schema": Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              },
              "uiSchema": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`NotificationEdit takes an endpoint 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationEdit
    endpoint={
      Object {
        "active": false,
        "filtersCount": 4,
        "id": 3,
        "name": "TEST Endpoint #3",
        "url": "http://endpoint3.com",
      }
    }
    fetchEndpoint={
      [MockFunction] {
        "calls": Array [
          Array [
            "1",
          ],
          Array [
            "1",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "endpointId": "1",
        },
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <PageHeader>
          <PageHeaderTitle
            title="Edit Notification"
          />
        </PageHeader>,
        <Connect(Main)>
          <Form
            ErrorList={[Function]}
            FieldTemplate={[Function]}
            className="pf-c-form"
            disabled={false}
            formData={
              Object {
                "active": false,
                "name": "TEST Endpoint #3",
                "url": "http://endpoint3.com",
              }
            }
            liveValidate={false}
            noHtml5Validate={false}
            noValidate={false}
            onChange={[Function]}
            safeRenderCompletion={false}
            schema={
              Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              }
            }
            uiSchema={Object {}}
          />
        </Connect(Main)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <PageHeaderTitle
            title="Edit Notification"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "title": "Edit Notification",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Form
            ErrorList={[Function]}
            FieldTemplate={[Function]}
            className="pf-c-form"
            disabled={false}
            formData={
              Object {
                "active": false,
                "name": "TEST Endpoint #3",
                "url": "http://endpoint3.com",
              }
            }
            liveValidate={false}
            noHtml5Validate={false}
            noValidate={false}
            onChange={[Function]}
            safeRenderCompletion={false}
            schema={
              Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              }
            }
            uiSchema={Object {}}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "ErrorList": [Function],
            "FieldTemplate": [Function],
            "className": "pf-c-form",
            "disabled": false,
            "formData": Object {
              "active": false,
              "name": "TEST Endpoint #3",
              "url": "http://endpoint3.com",
            },
            "liveValidate": false,
            "noHtml5Validate": false,
            "noValidate": false,
            "onChange": [Function],
            "safeRenderCompletion": false,
            "schema": Object {
              "properties": Object {
                "active": Object {
                  "default": true,
                  "title": "Active",
                  "type": "boolean",
                },
                "name": Object {
                  "default": "New notification endpoint name",
                  "title": "Name",
                  "type": "string",
                },
                "url": Object {
                  "default": "https://...",
                  "title": "URL",
                  "type": "string",
                },
              },
              "required": Array [
                "title",
              ],
              "title": "Edit Notifications",
              "type": "object",
            },
            "uiSchema": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <PageHeader>
            <PageHeaderTitle
              title="Edit Notification"
            />
          </PageHeader>,
          <Connect(Main)>
            <Form
              ErrorList={[Function]}
              FieldTemplate={[Function]}
              className="pf-c-form"
              disabled={false}
              formData={
                Object {
                  "active": false,
                  "name": "TEST Endpoint #3",
                  "url": "http://endpoint3.com",
                }
              }
              liveValidate={false}
              noHtml5Validate={false}
              noValidate={false}
              onChange={[Function]}
              safeRenderCompletion={false}
              schema={
                Object {
                  "properties": Object {
                    "active": Object {
                      "default": true,
                      "title": "Active",
                      "type": "boolean",
                    },
                    "name": Object {
                      "default": "New notification endpoint name",
                      "title": "Name",
                      "type": "string",
                    },
                    "url": Object {
                      "default": "https://...",
                      "title": "URL",
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "title",
                  ],
                  "title": "Edit Notifications",
                  "type": "object",
                }
              }
              uiSchema={Object {}}
            />
          </Connect(Main)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <PageHeaderTitle
              title="Edit Notification"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "title": "Edit Notification",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Form
              ErrorList={[Function]}
              FieldTemplate={[Function]}
              className="pf-c-form"
              disabled={false}
              formData={
                Object {
                  "active": false,
                  "name": "TEST Endpoint #3",
                  "url": "http://endpoint3.com",
                }
              }
              liveValidate={false}
              noHtml5Validate={false}
              noValidate={false}
              onChange={[Function]}
              safeRenderCompletion={false}
              schema={
                Object {
                  "properties": Object {
                    "active": Object {
                      "default": true,
                      "title": "Active",
                      "type": "boolean",
                    },
                    "name": Object {
                      "default": "New notification endpoint name",
                      "title": "Name",
                      "type": "string",
                    },
                    "url": Object {
                      "default": "https://...",
                      "title": "URL",
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "title",
                  ],
                  "title": "Edit Notifications",
                  "type": "object",
                }
              }
              uiSchema={Object {}}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "ErrorList": [Function],
              "FieldTemplate": [Function],
              "className": "pf-c-form",
              "disabled": false,
              "formData": Object {
                "active": false,
                "name": "TEST Endpoint #3",
                "url": "http://endpoint3.com",
              },
              "liveValidate": false,
              "noHtml5Validate": false,
              "noValidate": false,
              "onChange": [Function],
              "safeRenderCompletion": false,
              "schema": Object {
                "properties": Object {
                  "active": Object {
                    "default": true,
                    "title": "Active",
                    "type": "boolean",
                  },
                  "name": Object {
                    "default": "New notification endpoint name",
                    "title": "Name",
                    "type": "string",
                  },
                  "url": Object {
                    "default": "https://...",
                    "title": "URL",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                ],
                "title": "Edit Notifications",
                "type": "object",
              },
              "uiSchema": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
